[alternative] device: rework how we do struct device extern's

mbolivar-nordic opened the pull request:
This is an alternative to https://github.com/zephyrproject-rtos/zephyr/pull/47798.

Instead of introducing a new magic "DEVICE_ORD_MACRO" that callers have to override, we here add some new devicetree.h machinery for iterating over all the nodes in the tree.

That lets us replace device_extern.h with a foreach call in device.h in a way that should generalize analogously to emulators, without any tricky redefinitions.

mbolivar-nordic committed:
gen_defines: trivial rename

Generalize a function name as prep for extendings its scope.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
gen_defines: add helpers for iterating over the entire tree

Add two helper macros:

- DT_FOREACH_HELPER
- DT_FOREACH_OKAY_HELPER

These are internal generated macros which will be used to define
devicetree.h APIs for iterating over every node in the tree, and every
node in the tree with status "okay", respectively.

We can use these new APIs to implement the functionality in
device_extern.h natively from device.h without requiring a generated
header. It will also be useful for other purposes later on.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree.h: DT_FOREACH_NODE, DT_FOREACH_STATUS_OKAY_NODE

Add two new utility macros for iterating over the entire tree, along
with tests.

I have a use case for DT_FOREACH_STATUS_OKAY_NODE() right now, but I
think it makes sense to define both of them right away for
completeness.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: document internal macros for new for-each macros

Keep macros.bnf up to date with new generated macros.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
device.h: rework the device_extern.h mechanism

Now that we can iterate over all the status "okay" devicetree nodes
directly using <zephyr/devicetree.h>, use that along with a new helper
macro to pre-declare all possible struct devices.

Unlike the previous implementation, which declared devices even for
disabled nodes, we restrict the new mechanism to only declaring
devices for status "okay" nodes. This keeps drivers honest by not
pre-declaring devices for disabled nodes, which should not be
allocated in the first place.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
cmake: remove device_extern.h logic

This file is no longer used by device.h, so let's avoid spending time
generating the content formerly in it.

In order to leave a pointer in place for users who are expecting to
see it or are pulling it into their own builds, however, replace its
contents with an #error directive that tells them what happened. This
can be removed later on when we expect people are used to the
transition.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Twister errors look unrelated, but I found a typo in a comment, so I rebased and fixed it to see if this goes green now.

alevkoy commented on scripts/dts/gen_defines.py:
```diff
@@ -858,6 +858,16 @@ def write_global_macros(edt):
+
+    out_comment("Macros for iterating over all nodes and enabled nodes")
+    out_define(
+        "DT_FOREACH_HELPER(fn)",
```
I'm not sure how much readability difference this makes, but `out_dt_define` would work here.

alevkoy commented on include/zephyr/device.h:
```diff
@@ -950,12 +950,25 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+#define Z_MAYBE_DEVICE_DECLARE_INTERNAL(node_id) \
+	extern const struct device DEVICE_DT_NAME_GET(node_id);
+
+DT_FOREACH_STATUS_OKAY_NODE(Z_MAYBE_DEVICE_DECLARE_INTERNAL)
```
Shouldn't you `#include <zephyr/devicetree.h>` for this?

alevkoy reviewed (changes requested):
Only real request is to include what you use. Other comments are advisory.

gmarull commented on cmake/modules/dts.cmake:
```diff
@@ -231,11 +230,11 @@ if(SUPPORTS_DTS)
-    zephyr_file_copy(${DEVICE_EXTERN_H}.new ${DEVICE_EXTERN_H})
-    file(REMOVE ${ZEPHYR_DTS}.new ${DEVICETREE_UNFIXED_H}.new ${DEVICE_EXTERN_H}.new)
+    file(WRITE ${DEVICE_EXTERN_H}
+"#error The contents of this file are now implemented directly in zephyr/device.h.")
```
was there any case of direct `device_extern.h` inclusion? ie, was it a documented public header? Is it worth adding this?

mbolivar-nordic replied:
> was there any case of direct device_extern.h inclusion?

Not to my knowledge, but mcuboot used to include internal DT headers and the migration to the current generated macros broke the build there. I'm trying to be more helpful to users who are doing the wrong thing. Yes, they're wrong, but it's easy to point them in the right direction in this case.

>  ie, was it a documented public header?

No.

>  Is it worth adding this?

Yes, I think so, at **least** for 1 release, just as advice. Also consider that I just gave a talk about how the device extern mechanism works at ZDS that mentions this header explicitly. I think people who watched the talk will look for it and get confused without a little help here.

gmarull reviewed (commented):
approach lgtm, thanks for this PR! minor comments

mbolivar-nordic commented on cmake/modules/dts.cmake:
```diff
@@ -231,11 +230,11 @@ if(SUPPORTS_DTS)
     zephyr_file_copy(${DEVICETREE_UNFIXED_H}.new ${DEVICETREE_UNFIXED_H} ONLY_IF_DIFFERENT)
-    zephyr_file_copy(${DEVICE_EXTERN_H}.new ${DEVICE_EXTERN_H})
-    file(REMOVE ${ZEPHYR_DTS}.new ${DEVICETREE_UNFIXED_H}.new ${DEVICE_EXTERN_H}.new)
+    file(WRITE ${DEVICE_EXTERN_H}
```
> Are you planning to remove this file at a later date?

Yes. See the commit log.

mbolivar-nordic committed:
device.h: add missing include

This header makes extensive use of devicetree macros, but it does not
include zephyr/devicetree.h. Fix it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Rebased again and addressed comments. Networking related CI failures continue to appear unrelated.

mbolivar-nordic commented:
@stephanosio @rlubos any idea what's going on with the failing TCP tests here? I can't reproduce the failures locally and they all seem timing related, so I doubt they are caused by this PR.

rlubos commented:
> @stephanosio @rlubos any idea what's going on with the failing TCP tests here? I can't reproduce the failures locally and they all seem timing related, so I doubt they are caused by this PR.

The socket failures indeed seems to be timing related, I recall that `mps2_an385` already had timing problems in the past (the `select` suite) and the timeouts were loosened already. I suggest to rerun the tests, and if the problem persists to loosen the timings even further.

mbolivar-nordic commented:
Added DNM while I'm poking around with various fudge factors

mbolivar-nordic committed:
tests: net: socket: adjust fudge factors

Loosen some timing constraints in order to hack arond spurious
failures in upstream Zephyr's CI while working on an unrelated task.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
tests: kernel: common: adjust fudge factors

Loosen some timing constraints in order to hack arond spurious
failures in upstream Zephyr's CI while working on an unrelated task.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
The last CI run with adjusted fudge factors passed, so I've cleaned up the patches touching the unrelated tests and removed DNM.

@andyross can you please check `tests: kernel: common: adjust fudge factors` is not doing anything egregious?

@rlubos similarly can you give me an ACK on `tests: net: socket: adjust fudge factors`? I followed your advice to loosen the timings.

andyross commented on tests/kernel/common/src/clock.c:
```diff
@@ -35,7 +35,7 @@ K_TIMER_DEFINE(ktimer, duration_expire, stop_expire);
 
 #define DURATION 100
-#define LESS_DURATION 80
+#define LESS_DURATION 70
```
This looks fine.  On (non-icount/realtime) emulation platforms, a host scheduling quantum is on the order of 10ms.  So it used to be "2 host ticks", and now it's three.  That wasn't a lot of wiggle room on a loaded CI host, and all the test is doing is making sure that the timer hasn't fired yet anyway.

rlubos reviewed (approved):
Changes in net tests look good.
